WIP: 107 Nostr IoT (#814)
nostr-protocol/nips: 107.md

NIP-107
======

Nostr Internet of Things 
-----------------------------------

`draft` `optional` `author:benarc` `author:blackcoffeebtc`

> ## Ratioanale

Reviewer: ```suggestion
## Rationale
```

NIP-107
======

Nostr Internet of Things 
-----------------------------------

`draft` `optional` `author:benarc` `author:blackcoffeebtc`

## Ratioanale

The terms/conditions of IoT software/hardware is horrible. A user should be able to message a light or heating and ask it to turn on/off. Nostr gives the simple, permissionless development enviroment IoT needs.

All communication happens over [NIP04](https://github.com/nostr-protocol/nips/blob/master/04.md).

## Terms

- `user` - user operating IoT device with NOSTR key-pair and a client made specifically for IoT
- `device` - device to operate over Nostr with NOSTR key-pair, using a microcontroller client like <a href="https://github.com/lnbits/arduino-nostr">nostr-arduino</a>

## Nostr IoT Clients

### User

Where the `user` registers 'device'(s) and its keys, then updates the `device`(s).

### Device

The `device` uses a client like <a href="https://github.com/lnbits/arduino-nostr">nostr-arduino</a> to receive commands from the `user` or another `device`. 

> The `devie` can also push data such as senssor readings and updates. 

Reviewer: ```suggestion
The `device` can also push data such as sensor readings and updates.
```

## `User` OR `device` registering/updating a `user` OR `device`

A merchant can publish these events:
| Kind    |                  | Description                                                                                                   | NIP                                     |
|---------|------------------|---------------------------------------------------------------------------------------------------------------|-----------------------------------------|
| `4    ` | `direct_message` | Communicate between `device`(s) and `user`(s). The messages can be plain-text or JSON. | [NIP04](https://github.com/nostr-protocol/nips/blob/master/04.md) |

### Event: Register a `device`

**Event Content**:
```json
{
    "id": <String, UUID generated by the user.Sequential IDs (`0`, `1`, `2`...) are discouraged>,

>     "device_id": <String, UUID of the stall to which this device belong to>,

Reviewer: ```suggestion
    "device_id": <String, UUID of the device>,
```

    "name": <String, device name>,
    "description": <String (optional), device description>,
    "categories":[ <String (optional), device category, such as 'boiler'>],
    "users":[ [<String (optional), user public-key>]],

>     "images": <[String], array of image URLs, optional>

Reviewer: - is `"images"` a leftover?
```suggestion
```

}
```

Note:
 - `users`:
   - A list that can restrict access to public-keys, `device`(s) also has keys and can be a `user`.

### Event: Update a `device`

**Event Content**:
```json
{

>     "id": <String, UUID generated by the merchant.Sequential IDs (`0`, `1`, `2`...) are discouraged>,

Reviewer: ```suggestion
      "id": <String, UUID generated by the user. Sequential IDs (`0`, `1`, `2`...) are discouraged>,
```
- not sure if this `"id"` will be used in any way
